lib/url.js

```diff
@@ -681,44 +545,439 @@ Url.prototype.resolveObject = function(relative) {
+        hostnameEnd = i - 1;
+        break;
+      }
+      idna = true;
+      needsLowerCasing = true;
+    }
+    lastCh = ch;
+    charsAfterDot++;
+  }
+
+  var portEnd = i - 1;
+  hostnameEnd = portStart !== -1 ? portStart - 2 : hostnameEnd;
+
+  if (portStart !== -1)
+    hostEndsAtPortEnd = this._parsePort(str, portStart, portEnd);
+
+  // TODO(petkaantonov) This error is originally ignored:
+  // if (lastCh === 0x2E /*'.'*/)
+  //   hostnameEnd--
+
+  if (hostnameEnd + 1 !== start &&
+      hostnameEnd - hostNameStart <= 256) {
+    var hostname = str.slice(hostNameStart, hostnameEnd + 1);
+    if (needsLowerCasing)
+      hostname = hostname.toLowerCase();
+    if (idna)
+      hostname = punycode.toASCII(hostname);
+
+    this.hostname = hostname;
+    this.host =
+        this._isEmpty(this.port) ? hostname : hostname + ':' + this.port;
+  }
+
+  return (hostEndsAtPortEnd ? portEnd + 1 : portStart - 1);
 };
 
-// About 1.5x faster than the two-arg version of Array#splice().
-function spliceOne(list, index) {
-  for (var i = index, k = i + 1, n = list.length; k < n; i += 1, k += 1)
-    list[i] = list[k];
-  list.pop();
+Url.prototype._parsePath = function(str, start, end) {
+  var pathStart = start;
+  var pathEnd = end;
+  var escape = false;
+  var autoEscapeCharacters = _autoEscapeCharacters;
+
+  for (var i = start; i <= end; ++i) {
+    var ch = str.charCodeAt(i);
+
+    if (ch === 0x23 /*'#'*/) {
+      this._parseHash(str, i, end);
+      pathEnd = i - 1;
+      break;
+    } else if (ch === 0x3F /*'?'*/) {
+      this._parseQuery(str, i, end);
+      pathEnd = i - 1;
+      break;
+    } else if (!escape && autoEscapeCharacters[ch] === 1) {
+      escape = true;
+    }
+  }
+
+  if (pathStart > pathEnd) {
+    this.pathname = '/';
+    return;
+  }
+
+  var path;
+  if (escape)
+    path = getComponentEscaped(str, pathStart, pathEnd, false);
+  else
+    path = str.slice(pathStart, pathEnd + 1);
+
+  this.pathname = (this._prependSlash ? '/' + path : path);
+};
+
+Url.prototype._parseQuery = function(str, start, end) {
+  var queryStart = start;
+  var queryEnd = end;
+  var escape = false;
+  var autoEscapeCharacters = _autoEscapeCharacters;
+
+  for (var i = start; i <= end; ++i) {
+    var ch = str.charCodeAt(i);
+
+    if (ch === 0x23 /*'#'*/) {
+      this._parseHash(str, i, end);
+      queryEnd = i - 1;
+      break;
+    } else if (!escape && autoEscapeCharacters[ch] === 1) {
+      escape = true;
+    }
+  }
+
+  if (queryStart > queryEnd) {
+    this.search = '';
+    return;
+  }
+
+  var query = escape ?
+      getComponentEscaped(str, queryStart, queryEnd, true) :
+      str.slice(queryStart, queryEnd + 1);
+
+  this.search = query;
+};
+
+Url.prototype._parseHash = function(str, start, end) {
+  if (start > end) {
+    this.hash = '';
+    return;
+  }
+  this.hash = getComponentEscaped(str, start, end, true);
+};
+
+// This must be a method because this getting inlined is crucial.
```

Comment:
@domenic they do, but methods have polymorphic inline caches, whereas functions only have monomorphic caches. This function obviously accepts different types of argument, so V8 would deopt it if it were not a method.